csharp/src/Ice/Reference.cs

```diff
@@ -1820,5 +1518,292 @@ internal void Write(OutputStream ostr)
             }
             Debug.Assert(invocationTimeout != TimeSpan.Zero);
         }
+
+        private async Task<IncomingResponseFrame> PerformInvokeAsync(
+            OutgoingRequestFrame request,
+            bool oneway,
+            IProgress<bool>? progress,
+            bool releaseRequestAfterSent,
+            IInvocationObserver? observer,
+            CancellationToken cancel)
+        {
+            Connection? connection = _connection;
+            bool cached = false;
+            IReadOnlyList<Endpoint> endpoints;
+            List<Connector>? connectors = null;
+            OutgoingConnectionFactory connectionFactory = Communicator.OutgoingConnectionFactory;
+
+            // TODO replace IsConnectionCached with PreferExistingConnection
+            if ((connection == null || (!IsFixed && !connection.IsActive)) && IsConnectionCached)
+            {
+                // No cached connection, so now check if the connection factory has an existing connection that we
+                // can reuse, the connection factory will compute the reference endpoints and the endpoint
+                // connectors if required.
+                (connection, cached, endpoints, connectors) = await connectionFactory.GetConnectionAsync(
+                    this,
+                    cancel).ConfigureAwait(false);
+
+                _connection = connection;
+            }
+
+            int nextConnector = 0;
+            int attempt = 1;
+            bool triedAllConnectors = false;
+            List<Connector>? excludedConnectors = null;
+            IncomingResponseFrame? response = null;
+            Exception? exception = null;
+
+            while (true)
+            {
+                bool sent = false;
+                IChildInvocationObserver? childObserver = null;
+                try
+                {
+                    if (connection == null)
+                    {
+                        if (connectors == null)
+                        {
+                            // ComputeEndpointsAsync throws if it can't figure out the endpoints
+                            // TODO ComputeEndpointsAsync should return a integer indicating the locator cache version,
+                            // -1 indicates the locator was not contacted.
+                            (cached, endpoints) = await ComputeEndpointsAsync(cancel).ConfigureAwait(false);
```

Comment:
The plan is to update the locator cache in a follow-up PR, this will most probably return the tll, and also allow passing the tll to check if we need to refresh endpoints.